fixed bug id 7566, allow multiple sites for image graph generation

saleemkce opened the pull request:
This bug fix is for issue https://github.com/piwik/piwik/issues/7566 So, with this fix, now it's possible to give idSite=1 (single site image graph generation), idSite=1,2,3,4 (comma separated idSites) or idSite=all(list all sites the user has view permission).

During this fix, the only tradeoff is, I needed to create a new function **strokeAsImageData()** to send the generated image's content back to 'plugins/ImageGraph/StaticGraph.php' because current implementation causes the browser to output single PNG image. So, we refactor the code to just pass the image data back to the caller so that it could be utilized in HTML page with N no. of images for each image graph generation.

It's still not better approach because this new function is added in pChar library file. We need to inherit the code so that when pChart library is updated, we don't lose our new **strokeAsImageData()** function. Please go through the code and give suggests, we can improve further.

For time being, use my fix and play with this new multi image graph page.
  http://localhost.piwik/index.php?module=API&method=ImageGraph.get&idSite=all&apiModule=VisitsSummary&apiAction=get&token_auth=anonymous&graphType=evolution&period=day&date=previous30&width=500&height=250

saleemkce committed:
fixed bug id 7566, allow multiple sites for image graph generation

saleemkce commented:
Current newly implemented page looks like this:

![multiimagegraph](https://cloud.githubusercontent.com/assets/6330544/7101630/d325b54c-e07f-11e4-88ce-9c26a4757bc7.png)

saleemkce commented:
hi all, what is the status for this commit? If you have feedback, please comment. I will do everything to get it done. Do you find issues with it or could it be moved to master?

saleemkce commented:
Here is a bar graph representation for multiple websites.

![multiimagebargraph](https://cloud.githubusercontent.com/assets/6330544/7257620/a5b4835c-e874-11e4-97c4-8e2acb8ab7d9.png)


Oh, how nice it looks in verticalBar, horizontalBar and as Pie. I cannot wait to see how long it will take before this commit is approved and moved to Piwik codebase. But everyone is eager to see it done sooner. And me too.

mattab commented:
Hi @saleemkce 

Thanks for the pull request! The diff looks quite big so it may take a while before we get a chance to review and consider it. Also we'd need to add some tests for this (see also:  Enable static PNG image graphs during Continuous Integration #6864 )

saleemkce commented:
@mattab Okay. Very glad t to know. Moreover, if you take a look at this http://forum.piwik.org/read.php?9,126057, it would be helpful.

saleemkce committed:
Merge branch 'master' of https://github.com/piwik/piwik

saleemkce committed:
Merge branch 'master' of https://github.com/piwik/piwik

tsteur commented on libs/pChart2.1.3/class/pImage.class.php:
```diff
@@ -166,6 +166,17 @@ function stroke($BrowserExpire=FALSE)
     }
+    
+   /* Updated code to send image as encoded data */
+   function strokeAsImageData()
```
Would it be possible to the `ob_start` etc the StaticGraph class eg in a private method `strokeAsImageData`? Would be good if it was possible. Otherwise we always need to make sure to apply this patch when updating the pchart library etc.

saleemkce replied:
Hi Thomas, I am very unsure about this. Could you put some sample code and clarify me how we could get it done better without getting affected by an update from pchart's library?

tsteur replied:
I meant can you do something like 

```php
public function getPngImageData()
{
   return 'data:image/png;base64,'.base64_encode($this->strokeAsImageData());
}

private function strokeAsImageData() 
{
     ob_start();
     $this->pImage->stroke()
     $imagedata = ob_get_contents();
     ob_end_clean();
     return $imagedata;
}
```

tsteur commented on plugins/ImageGraph/API.php:
```diff
@@ -129,404 +130,453 @@ public function get(
         $segment = false
     )
     {
-        Piwik::checkUserHasViewAccess($idSite);
```
Here you could use `Piwik::checkUserHasViewAccess($idSites)`

tsteur commented on plugins/ImageGraph/API.php:
```diff
@@ -129,404 +130,453 @@ public function get(
             }
+        } else {
+            // empty or invalid idSite has been already handled somewhere
+            exit;
```
We usually don't to `exit` in our code. Can we instead throw an exception or return?

saleemkce replied:
Yes, I would remove exit else loop from here. It already throws exception before it reaches this point.

tsteur commented on plugins/ImageGraph/API.php:
```diff
@@ -129,404 +130,453 @@ public function get(
-            $apiParameters = array();
-            if (!empty($idGoal)) {
-                $apiParameters = array('idGoal' => $idGoal);
+        $idSiteArr = array();
```
Here you could use `$idSiteArr = Site::getIdSitesFromIdSitesString($idSite)`. This handles eg the case of `idSite=all` and does an `explode(',', $idSite)` etc

saleemkce replied:
```
<?php
Site::getIdSitesFromIdSitesString($idSite);
```
That's great. I will make use of it. Thanks for notifying this.

tsteur replied:
No worries, it is basically impossible to find if you don't know it

tsteur commented:
The diff is quite big indeed. I wonder if it was possible to make the code simpler by leaving the behaviour of `ImageGraph.get` more or less as it is. Then introduce a new method like `ImageGraph.getForSites` which might use `ImageGraph.get($idSite)` for each given site and puts the graph together. Or maybe we could extract some code from `ImageGraph.get` into methods and reuse them in `getForSites` (feel free to change the name if you do it).

tsteur commented on plugins/ImageGraph/API.php:
```diff
@@ -129,404 +130,453 @@ public function get(
+        for ($w = 0; $w < $idSiteArrCount; $w++)
+        {
+            $idSite = $idSiteArr[$w];
+            Piwik::checkUserHasViewAccess($idSite);
```
Would be no longer needed if doing `Piwik::checkUserHasViewAccess($idSite)` in the beginning

saleemkce replied:
Let's consider we have

```
<?php
$idSiteArr = array(1,2,3,4,5);
```

Is it possible to pass an array of idSites like this outside the for loop?

for example,
Piwik::checkUserHasViewAccess($idSiteArr);

I had used like this **"Piwik::checkUserHasViewAccess($idSite);"** because I thought this function could check idSite one at a time. Could it take array as given above?  

Plus what will happen if user doesn't have view access only for idSide=3. Will it output other 4 image graphs or just stop after outputting first 2?

saleemkce replied:
>  "The diff is quite big indeed. I wonder if it was possible to make the code ..."

Hi Thomas, there're actually very few changes added to existing code. **Not big** as you said. Github diff tool may make it seem too big. What I did in **ImageGraph.get** function is just put a **for loop** around the statements inside the function and storing the image graph outputs to an array variable so that it could be used later. It's that simple. It doesn't alter the output behavior of any other function parameters supplied to this function. If you think that it really affects the existing behavior, then we may move it to new function. Otherwise, I don't find any reason to reuse the code and write the new functionality in a new function. Moreover, that would increase the program LOC drastically which is really not useful. Please apply this patch in you app and play around with this a few more times. You may find something more constructive. What do you think?

tsteur replied:
`Piwik::checkUserHasViewAccess($idSite);` you can do this at the very beginning, then you don't have to do it in a for loop. Followed by this you can do `$idSiteArr = Site::getIdSitesFromIdSitesString($idSite)` as this will exactly do what you want and you can remove all the custom logic re `$idSiteArr`.

Yes the diff is smaller now

saleemkce committed:
Merge branch 'master' of https://github.com/piwik/piwik

mattab commented:
> The diff is quite big

Btw I just noticed if we look at the diff without whitespace (`&w=1`) it looks smaller: https://github.com/piwik/piwik/pull/7660/files?w=1

tsteur commented on plugins/ImageGraph/API.php:
```diff
@@ -129,404 +130,453 @@ public function get(
+     * @param string $imageGraphCount Number of img tags available for output
+     * Exits after outputting HTML content
+     */
+    public function renderImageGraph($pngImgArr, $imageGraphCount)
```
This should be a private method. Otherwise it will be added to the API documentation as an API

tsteur commented on plugins/ImageGraph/API.php:
```diff
@@ -129,404 +130,453 @@ public function get(
+        $imageGraphCount = count($pngImgArr);
+        if ($imageGraphCount >= 1) {
+            $this->renderImageGraph($pngImgArr, $imageGraphCount);
+        }
```
I'm not quite sure. Shouldn't there be a 
```php
else {
    $graph->sendToBrowser();
    // or something like $pngImgArr[0]['source']->sendToBrowser();
}
```

somewhere in case only one graph is given?

saleemkce replied:
> if ($imageGraphCount >= 1) {

Line 599 above. This line handles even if there is only one graph so we don't need else loop here.

tsteur replied:
Sure, doesn't this return an HTML page then when there is only one site (and it was an image before)? It is important that it still returns an image only for one site as it is for example displayed in Android and iOS app. An HTML page would not work there. I might be wrong :)

saleemkce replied:
yes, you are right. I admit. We should discuss about this further.

tsteur commented on plugins/ImageGraph/API.php:
```diff
@@ -129,404 +130,453 @@ public function get(
+     */
+    public function renderImageGraph($pngImgArr, $imageGraphCount)
+    {
+        header('Content-Type: text/html; charset=utf-8');
```
ImageGraph would ideally return an image and not an HTML page. That's what ImageGraph API is about. Do you think it would be possible to render this - somehow - as an image instead of HTML page? I'm not so much into this library so I do not know whether it is a big deal. If it is not easily possible, should be maybe introduce a new output type? or move it to a new API method?

cc @mattab what are your thoughts?

saleemkce committed:
Merge branch 'master' of https://github.com/piwik/piwik

saleemkce committed:
some important suggestions implemented.

saleemkce committed:
Fix makes use of StaticGraph to achieve the same.

Thanks to Thomas' brilliant suggestion which helped me achieve this without touching pchart's library. So, I remove this function from here and do the functionality at plugins/ImageGraph/SaticGraph.php file.

saleemkce committed:
Fix from a key discussion about pChart reuse.

saleemkce committed:
Merge branch 'bugfix-multisite-graph' of https://github.com/saleemkce/piwik into bugfix-multisite-graph

saleemkce commented:
@mattab  @tsteur Thanks a lot, guys. All your suggestions successfully implemented. I have tested all new changes with 2 sites in my local machine. Everything is okay and looks nice as before. Latest changes here: https://github.com/piwik/piwik/pull/7660/files?w=1 Please notify if anything is wrong.

mattab commented:
Dear @saleemkce 

Thanks for the Pull request. At this stage, we will not merge it as we think we don't need this feature in Piwik. For example, it's quite simple for someone to include graphs from multiple websites by creating a little PHP script that loops through sites and then generates the image. Therefore we prefer not to add complexity in this part of the code. Thanks anyway for your PR and we hope you will keep contributing in the future :-)
